File: backend/models/promotion.py
from datetime import datetime, timezone
from enum import Enum
from typing import List, Optional
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator

from models.common import clean_optional_text


class PromoType(str, Enum):
    PERCENTAGE = "percentage"
    FIXED_AMOUNT = "fixed_amount"
    FREE_DELIVERY = "free_delivery"
    EXPRESS_UPGRADE = "express_upgrade"


class PromoTarget(str, Enum):
    ALL = "all"
    FIRST_DELIVERY = "first_delivery"
    TIER_SILVER = "tier_silver"
    TIER_GOLD = "tier_gold"
    DELIVERY_MODE = "delivery_mode"


class PromotionCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=120)
    description: str = Field(default="", max_length=1000)
    promo_type: PromoType
    value: float = Field(default=0.0, ge=0, le=1_000_000)
    target: PromoTarget = PromoTarget.ALL
    delivery_mode: Optional[str] = Field(default=None, max_length=80)
    target_user_ids: Optional[List[str]] = Field(default=None, max_length=1000)
    min_amount: Optional[float] = Field(default=None, ge=0)
    max_uses_total: Optional[int] = Field(default=None, ge=1)
    max_uses_per_user: int = Field(default=1, ge=1, le=100000)
    promo_code: Optional[str] = Field(default=None, max_length=40)
    start_date: datetime
    end_date: datetime
    is_active: bool = True

    @field_validator("title", "description", "delivery_mode", "promo_code")
    @classmethod
    def normalize_text_fields(cls, value: Optional[str]) -> Optional[str]:
        return clean_optional_text(value)


class Promotion(PromotionCreate):
    promo_id: str = Field(default_factory=lambda: f"promo_{uuid4().hex[:12]}")
    uses_count: int = 0
    created_by: str = ""
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class PromoUse(BaseModel):
    use_id: str
    promo_id: str
    user_id: str
    parcel_id: str
    created_at: datetime

File: mobile/lib/core/models/promotion.dart
class Promotion {
  const Promotion({
    required this.promoId,
    required this.title,
    required this.description,
    required this.promoType,   // "percentage" | "fixed_amount" | "free_delivery" | "express_upgrade"
    required this.value,
    required this.target,
    required this.startDate,
    required this.endDate,
    required this.isActive,
    required this.usesCount,
    this.promoCode,
    this.deliveryMode,
    this.minAmount,
    this.maxUsesTotal,
    this.maxUsesPerUser = 1,
  });

  final String   promoId;
  final String   title;
  final String   description;
  final String   promoType;
  final double   value;
  final String   target;
  final DateTime startDate;
  final DateTime endDate;
  final bool     isActive;
  final int      usesCount;
  final String?  promoCode;
  final String?  deliveryMode;
  final double?  minAmount;
  final int?     maxUsesTotal;
  final int      maxUsesPerUser;

  factory Promotion.fromJson(Map<String, dynamic> j) => Promotion(
        promoId:        j['promo_id'] as String,
        title:          j['title'] as String,
        description:    j['description'] as String? ?? '',
        promoType:      j['promo_type'] as String,
        value:          (j['value'] as num?)?.toDouble() ?? 0,
        target:         j['target'] as String? ?? 'all',
        startDate:      DateTime.parse(j['start_date'] as String),
        endDate:        DateTime.parse(j['end_date'] as String),
        isActive:       j['is_active'] as bool? ?? true,
        usesCount:      j['uses_count'] as int? ?? 0,
        promoCode:      j['promo_code'] as String?,
        deliveryMode:   j['delivery_mode'] as String?,
        minAmount:      (j['min_amount'] as num?)?.toDouble(),
        maxUsesTotal:   j['max_uses_total'] as int?,
        maxUsesPerUser: j['max_uses_per_user'] as int? ?? 1,
      );

  String get typeLabel => switch (promoType) {
        'percentage'      => '-${value.toStringAsFixed(0)}%',
        'fixed_amount'    => '-${value.toStringAsFixed(0)} XOF',
        'free_delivery'   => 'Livraison gratuite',
        'express_upgrade' => 'Express offert',
        _                 => '',
      };

  bool get isCurrentlyActive {
    final now = DateTime.now();
    return isActive && now.isAfter(startDate) && now.isBefore(endDate);
  }
}
